node/src/Pos/Wallet/Web/Pending/Functions.hs

```diff
@@ -0,0 +1,73 @@
+module Pos.Wallet.Web.Pending.Functions
+    ( reevaluateApplyingPtxs
+    ) where
+
+import           Universum
+
+import qualified Data.HashMap.Strict          as HM
+import           Control.Lens                 (has, each)
+
+import           Pos.Core.Configuration       (HasConfiguration)
+import           Pos.Core.Slotting            (flatSlotId)
+import           Pos.Core.Types               (ChainDifficulty, FlatSlotId, SlotId)
+import           Pos.Txp.Core                 (TxId, topsortTxs)
+import           Pos.Txp.Toil.Class           (MonadUtxoRead)
+import           Pos.Wallet.Web.Pending.Types (PendingTx (..), PtxCondition (..),
+                                               PtxSubmitTiming (..), pstNextDelay,
+                                               pstNextSlot, ptxPeerAck, ptxSubmitTiming, ptxCond, _PtxApplying)
+
+
+-- | Reevaluate all pending txs (in 'PtxApplying' state):
+--
+--   * 'PtxWontApply' for ones that can't be applied
+--   * 'PtxInNewestBlocks' for ones that are already
+--     in the blockchain (at any depth)
+--   * 'PtxApplying' to ones that can be applied either directly or after
+--     applying some other transactions
+--
+-- TODO: do we need to do anything special to ensure that the UTXO we're
+-- shown doesn't include outputs from failed txs or mempool txs? I'm not
+-- sure
+reevaluateApplyingPtxs
+    :: (HasConfiguration, MonadUtxoRead m)
+    => HashMap TxId PendingTx
+    -> ChainDifficulty
+    -> m (HashMap TxId PendingTx)
+reevaluateApplyingPtxs ptxs curDifficulty = do
+    -- Get all 'PtxApplying' transactions
+    let applying, notApplying :: HashMap TxId PendingTx
+        applying    = HM.filter (has (ptxCond . _PtxApplying)) ptxs
+        notApplying = HM.filter (not . has (ptxCond . _PtxApplying)) ptxs
+    -- Find those that are already in blocks; they'll be marked as
+    -- 'PtxInNewestBlocks'
+    let (present, missing) = undefined  -- HM.partition ....... applying
+    -- Topsort the rest ('missing'), go one by one and either add to UTXO or
+    -- mark as 'PtxWontApply'
+    (missingInvalid, missingValid) <- runToilT $ do
+        let sorted = topsortTxs undefined missing
+        -- TODO: is it okay that we don't do 'verifyGState' here?
+        fmap partitionEithers $ forM sorted $ \tx -> do
+            let vtc = VTxContext True     -- TODO: is it necessarily 'True'?
+            runExceptT (verifyTxUtxo vtc tx) >>= \case
```

Comment:
I think you should use `txProcessTransaction` here, otherwise if one tx depends on another one, the second one will fail (because the first one won't be applied).

Reply:
It will be applied, I call `applyTxToUtxo` if verification succeeds. I don't know if I can use `txProcessTransaction` here because it looks like it also adds the transaction to the mempool, and writes logs, and does whatever other horrible things. Maybe they aren't horrible. Or maybe they can be avoided. I don't know.

Reply:
Ah, okay, you use `applyTxToUtxo` indeed.
I think those things aren't horrible, but this code makes sense to me now. Looks like we indeed don't need to `verifyGState` here, so it should be fine.